csharp/src/Ice/Discovery/Locator.cs

```diff
@@ -136,31 +140,67 @@ internal class Locator : IAsyncLocator
                     adapterId.Length > 0 ? ImmutableArray.Create(adapterId) : ImmutableArray<string>.Empty);
         }
 
-        internal Locator(ILocatorRegistryPrx registry, ILookupPrx lookup, ObjectAdapter replyAdapter, string pluginName)
+        internal static ILocatorPrx Initialize(Communicator communicator) => new Locator(communicator)._proxy;
+
+        private Locator(Communicator communicator)
         {
-            _pluginName = pluginName;
-            _registry = registry;
-            _replyAdapter = replyAdapter;
+            const string defaultIPv4Endpoint = "udp -h 239.255.0.1 -p 4061";
+            const string defaultIPv6Endpoint = "udp -h \"ff15::1\" -p 4061";
 
-            Communicator communicator = replyAdapter.Communicator;
+            if (communicator.GetProperty("Ice.Discovery.Multicast.Endpoints") == null)
+            {
+                communicator.SetProperty("Ice.Discovery.Multicast.Endpoints",
+                                          $"{defaultIPv4Endpoint}:{defaultIPv6Endpoint}");
+            }
+
+            string? lookupEndpoints = communicator.GetProperty("Ice.Discovery.Lookup");
+            if (lookupEndpoints == null)
+            {
+                List<string> endpoints = new ();
+                List<string> ipv4Interfaces = Network.GetInterfacesForMulticast("0.0.0.0", Network.EnableIPv4);
+                List<string> ipv6Interfaces = Network.GetInterfacesForMulticast("::0", Network.EnableIPv6);
+
+                endpoints.AddRange(ipv4Interfaces.Select(i => $"{defaultIPv4Endpoint} --interface \"{i}\""));
+                endpoints.AddRange(ipv6Interfaces.Select(i => $"{defaultIPv6Endpoint} --interface \"{i}\""));
+
+                lookupEndpoints = string.Join(":", endpoints);
+            }
 
-            _timeout = communicator.GetPropertyAsTimeSpan($"{_pluginName}.Timeout") ?? TimeSpan.FromMilliseconds(300);
+            if (communicator.GetProperty("Ice.Discovery.Reply.Endpoints") == null)
+            {
+                communicator.SetProperty("Ice.Discovery.Reply.Endpoints", "udp -h \"::0\" -p 0");
+            }
+            communicator.SetProperty("Ice.Discovery.Reply.ProxyOptions", "-d"); // create datagram proxies
+
+            if (communicator.GetProperty("Ice.Discovery.Locator.Endpoints") == null)
```

Comment:
Why are we still checking for this if this is supposed to be a local adapter now? I'm not sure it's still useful to set the adapter ID as well, can't we just create an anonymous OA and rely on a well-known colocated proxy for the locator?

Reply:
Yes, fixed.